lib/Sema/CSSimplify.cpp

```diff
@@ -4552,36 +4552,28 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
   std::tie(fromObjectType, fromUnwraps) = getObjectTypeAndUnwraps(fromType);
   std::tie(toObjectType, toUnwraps) = getObjectTypeAndUnwraps(toType);
 
-  // Since equality is symmetric and it decays into a `Bind`, eagerly
-  // unwrapping optionals from either side might be incorrect since
-  // there is not enough information about what is expected e.g.
-  // `Int?? equal T0?` just like `T0? equal Int??` allows `T0` to be
-  // bound to `Int?` and there is no need to unwrap. Solver has to wait
-  // until more information becomes available about what `T0` is expected
-  // to be before taking action.
-  if (matchKind == ConstraintKind::Equal &&
-      (fromObjectType->is<TypeVariableType>() ||
-       toObjectType->is<TypeVariableType>())) {
+  // This failure is comming from a binding of a type variable to a fixed type
+  // e.g. $T0? bind Int
+  auto isBindingToFixedType = matchKind == ConstraintKind::Bind &&
+                              fromObjectType->is<TypeVariableType>() &&
+                              !toObjectType->is<TypeVariableType>();
+  // If there are type variable on either side eagerly unwrapping optionals from
+  // either side might be incorrect since there is not enough information about
+  // what is expected and there is no need to unwrap. Solver has to wait until
+  // more information becomes available about what those variables are expected
+  // to be before taking action. Except for when binding to a fixed type because
+  // we fail but fromUnwraps <= toUnwraps and fromObjectType can be bound to
+  // this concrete type, we know a force optional is the correct fix.
+  if (!isBindingToFixedType && (fromObjectType->is<TypeVariableType>() ||
+                                toObjectType->is<TypeVariableType>())) {
```

Comment:
So how about this idea - let's remove `isBindingToFixedType` and simplify this check into:

```
if (fromObjectType->is<TypeVariableType>() || toObjectType->is<TypeVariableType>())
```

and at the same time expand return type of this method from simple true/false into a `TypeMatchResult` - in case when this check triggers it should return `amibiguous` and all the call sites in `repairFailures` would `break` if it does, this way we'd make sure that there are no fixes would be recorded in this case which I believe should also fix the case with `?.` syntax you mentioned.

Reply:
Right, sounds good. I'll take a look on that approach in more detail soon. Initially with a quick change that cause that case specifically failed and end up without a fix and default diagnostic, but that may be just me messing something up. Let's see... thank you for the idea!

Reply:
Unfortunately I couldn't make it work, trying delay that way for this case in specific only end leaving the expression without a diagnostic because that would never comeback to that failure a re-attempt repair with fixed types assigned.